### PR TITLE
Implement plugin/extension to allow converting structured data (YAML) into Metanorma content

### PR DESCRIPTION
**Pull request by w00lf**
#195 

First draft Asciidoctor::Standoc::Yaml2TextPreprocessor - preprocess yaml2text blocks
@ronaldtse @opoudjis this is the first draft, i decide to implement this feature with help of `Asciidoctor::Extensions::Preprocessor`, can you please look and tell if i am moving in the right direction?

### CODE REVIEW

**Comment by ronaldtse** [comment]
@w00lf looks good to me, I'd like to more examples implemented as specs, and since the evaluation process can be nested, we will want to make sure that works correctly.

**Comment by opoudjis** [comment]
So far, yes it looks good.

The predecessor to this is unusable because it had zero unit testing and zero documentation; what you are doing is already superior.

You are doing line preprocessing instead of using the (poorly undocumented) Asciidoc API to build up constructs; that's fine, but you might need to be careful for that reason about edge cases.

**Comment by w00lf** [comment]
@ronaldtse @opoudjis done with all edge cases, please review

**Comment by ronaldtse** [comment]
@w00lf sorry for the late reply. This is a document that can use this structure:

https://gitlab.com/calconnect/tc-localization/csd-transcription-systems/

Can you refactor the document's `script/split_codes.rb` (https://gitlab.com/calconnect/tc-localization/csd-transcription-systems/-/blob/master/scripts/split_codes.rb) to utilize this feature as a demonstration?

i.e.

Before:
codes.csv => codes.adoc

Now:
codes.csv => codes.yaml => ingested by `yaml2text`

Thanks!

@w00lf ping me for access.

**Comment by w00lf** [comment]
> @w00lf sorry for the late reply. This is a document that can use this structure:
> 
> https://gitlab.com/calconnect/tc-localization/csd-transcription-systems/
> 
> Can you refactor the document's `script/split_codes.rb` (https://gitlab.com/calconnect/tc-localization/csd-transcription-systems/-/blob/master/scripts/split_codes.rb) to utilize this feature as a demonstration?
> 
> i.e.
> 
> Before:
> codes.csv => codes.adoc
> 
> Now:
> codes.csv => codes.yaml => ingested by `yaml2text`
> 
> Thanks!
> 
> @w00lf ping me for access.

Tested with file generated from refactored https://gitlab.com/calconnect/tc-localization/csd-transcription-systems/-/blob/master/scripts/split_codes.rb - see spec/asciidoctor-standoc/macros_yaml2text_spec.rb last test case. Generation process takes almost 15 seconds to complete because of the size of codes.csv file and result html

**Comment by ronaldtse** [comment]
@ribose-jeffreylau can you help check that this PR addresses the original specification? Thanks.

@w00lf we also need to add documentation for this feature on metanorma.com , thanks!

File: spec/asciidoctor-standoc/macros_yaml2text_spec.rb [hunk 1]
```diff
@@ -0,0 +1,467 @@
+require 'spec_helper'
+
+RSpec.describe Asciidoctor::Standoc::Yaml2TextPreprocessor do
+  describe '#process' do
+    let(:example_file) { 'example.yml' }
+
+    before do
+      if defined?(example_yaml_content)
+        File.new(example_file, 'w').tap { |n| n.puts(example_yaml_content) }.close
```
Comment (ribose-jeffreylau) [inline spec/asciidoctor-standoc/macros_yaml2text_spec.rb:9]: Would the following be more idiomatic?

```ruby
File.open(example_file, 'w') { |n| n.puts(example_yaml_content) }
```

Comment (w00lf) [reply spec/asciidoctor-standoc/macros_yaml2text_spec.rb]: Done, changed

**Comment by w00lf** [comment]
> Sorry, please address the comment I mentioned earlier.

@ribose-jeffreylau changed code, please review. Also, @ronaldtse @CAMOBAP795 i am geting LaTeXML error during ubuntu build, did our ci repo changed recently? Before these specs passed ok.

**Comment by CAMOBAP795** [comment]
@w00lf looks like not related to ci configuration changes, will be fixed in scope of https://github.com/metanorma/metanorma-standoc/issues/244

**Comment by opoudjis** [comment]
The method lengths are longer than they should be (we're forced to 10), but code looks perfectlhy tidy.

**Pull request merged**